### ISSUE TITLE
Documentation on parameters to change to improve model

### ISSUE THREAD

**Issue by lldelisle** [aertslab/CREsted#43]
### Description of feature

Dear CREsted dev,
Thank you very much for this tool and the step by step tutorial. I managed to generate a model however, when I tested, the result was disappointing:
```bash
2024-10-14T09:48:04.350399+0200 INFO Test concordance_correlation_coefficient: 0.4360
2024-10-14T09:48:04.350597+0200 INFO Test cosine_similarity: 0.8162
2024-10-14T09:48:04.350651+0200 INFO Test loss: -0.3639
2024-10-14T09:48:04.350690+0200 INFO Test mean_absolute_error: 0.0017
2024-10-14T09:48:04.350726+0200 INFO Test mean_squared_error: 0.0000
2024-10-14T09:48:04.350766+0200 INFO Test pearson_correlation: 0.4482
2024-10-14T09:48:04.350800+0200 INFO Test pearson_correlation_log: 0.6173
2024-10-14T09:48:04.350832+0200 INFO Test zero_penalty_metric: 29.8548
```
I feel like a correlation coefficient of 0.43 is much lower than what you have in your example (0.7149).

- Do you have a threshold value above which one the model is 'usable'? My goal is to get the contribution score on defined region.

- Which are the parameters that you would recommend to play with to try to improve this?

- Should we change the seed when attributing the train/test/validation when we create a new model?

Thank you very much.

**Comment by nkempynck** [comment]
Hi Lucille

A correlation of 0.43 indeed seems quite low. Could you provide more information on the model training?

For example:

- How many classes are in the model?
- How many regions are you training on?
- What training configuration are you using? Which loss function?
- Is this in the training on all peaks, or on the finetuning? If it is during finetuning, maybe the base model you chose was overfitting.

If you have an overfitting issue, the issue could be related to the size of your model. But again, then it would be useful to know how many classes and regions you have.

Cheers
Niklas

**Comment by lldelisle** [comment]
Sure, 
I have 8 classes. On the full peaks, here are the numbers for each split category:
```
train    112616
val       14077
test      14077
```
After filtering: `INFO After specificity filtering, kept 22906 out of 140770 regions.`

The model was generated in 2 steps. First step:
```python
datamodule = crested.tl.data.AnnDataModule(
    adata,
    genome_file="/home/ldelisle/genomes/fasta/mm10.fa",
    batch_size=64,
    max_stochastic_shift=3,  # optional augmentation
    always_reverse_complement=True,  # default True. Will double the effective size of the training dataset.
)
model_architecture = crested.tl.zoo.chrombpnet(seq_len=2114, num_classes=len(list(adata.obs_names)))
optimizer = keras.optimizers.Adam(learning_rate=1e-3)
loss = crested.tl.losses.CosineMSELogLoss(max_weight=100)
metrics = [
    keras.metrics.MeanAbsoluteError(),
    keras.metrics.MeanSquaredError(),
    keras.metrics.CosineSimilarity(axis=1),
    crested.tl.metrics.PearsonCorrelation(),
    crested.tl.metrics.ConcordanceCorrelationCoefficient(),
    crested.tl.metrics.PearsonCorrelationLog(),
    crested.tl.metrics.ZeroPenaltyMetric(),
]

alternative_config = crested.tl.TaskConfig(optimizer, loss, metrics)
trainer = crested.tl.Crested(
    data=datamodule,
    model=model_architecture,
    config=alternative_config,
    # config=config,
    project_name="test1_wt",  # change to your liking
    run_name="dyn_log_loss",  # change to your liking
    logger=None,  # or 'wandb', 'tensorboard'
)
trainer.fit(epochs=60)
```
This stopped after 31 Epoch.

The second step is on the filtered:
```
# Loading the pretrained model on all consensuspeaks and finetuning with lower learning rate
datamodule = crested.tl.data.AnnDataModule(
    adata,
    genome_file=genome_file,
    batch_size=64,  # Recommended to go for a smaller batch size than in the pretrained model
    max_stochastic_shift=3,
    always_reverse_complement=True,
)
# First load the pretrained model on all peaks
import keras
# I don't know if the model when to the end seems like it ended after epoch 31.
model_architecture = keras.models.load_model(
    "test1_wt/dyn_log_loss/checkpoints/21.keras", compile=False
)

# Use the same config you used for the pretrained model. EXCEPT THE LEARNING RATE, make sure that is lower than it was on the epoch you select the model from.
import keras

optimizer = keras.optimizers.Adam(learning_rate=1e-4)  # Lower LR!
loss = crested.tl.losses.CosineMSELogLoss(max_weight=100)
metrics = [
    keras.metrics.MeanAbsoluteError(),
    keras.metrics.MeanSquaredError(),
    keras.metrics.CosineSimilarity(axis=1),
    crested.tl.metrics.PearsonCorrelation(),
    crested.tl.metrics.ConcordanceCorrelationCoefficient(),
    crested.tl.metrics.PearsonCorrelationLog(),
    crested.tl.metrics.ZeroPenaltyMetric(),
]

alternative_config = crested.tl.TaskConfig(optimizer, loss, metrics)
print(alternative_config)

# setup the trainer
trainer = crested.tl.Crested(
    data=datamodule,
    model=model_architecture,
    config=alternative_config,
    project_name="test1_wt",  # change to your liking
    run_name="dyn_log_loss_FT",  # change to your liking
    logger=None,  # or 'wandb', 'tensorboard'
)

trainer.fit(epochs=60)
```
This one stopped after 32 Epoch.

Tell me if I can provide more info.

**Comment by nkempynck** [comment]
I think I see a potential issue. Do you run the specificity filtering before training your basemodel? This will filter out too many regions probably. You should train the basemodel on all peaks, and before the finetuning, do the specificity filtering.

**Comment by lldelisle** [comment]
No, sorry, I first generated 2 anndata, one for the whole set of peaks and one for the filtered. In the first part I use the whole set of peaks and in the second I use the filtered.

**Comment by lldelisle** [comment]
On the first part I get in the stderr:
```
100%|██████████| 112616/112616 [00:02<00:00, 44572.14it/s]                                                                                                  
100%|██████████| 14077/14077 [00:00<00:00, 46187.85it/s]  
```
On the second part, I get in the stderr:
```
100%|██████████| 18420/18420 [00:32<00:00, 565.63it/s] 
100%|██████████| 2253/2253 [00:03<00:00, 746.71it/s]
```

**Comment by nkempynck** [comment]
Ok. On first sight you seem to follow the steps correctly. Maybe the number of peaks is on the lower end, but that does not mean it shouldn't work. You can always try to add more peaks as background, to have more training data. The pearson correlation metric does not tell the whole story however. What is the val cosine similarity, and the val pearson log? Do you have the feeling the model does not work at all if you do example predictions on random test set regions? Can you maybe make the heatmaps of log correlation between predictions vs peaks and peaks vs peaks (in notebook) and share them? 

I would also recommend to also use Weights and Biases, and use the 'wandb' option for the 'logger' parameter for the crested class, in that way you can have a graphical interpretation of the model's performance over the different epochs. Bes

**Comment by lldelisle** [comment]
>  the number of peaks is on the lower end

I will rerun with more peaks and keep you updated.

> What is the val cosine similarity, and the val pearson log? 

```
2024-10-14T09:48:04.350399+0200 INFO Test concordance_correlation_coefficient: 0.4360
2024-10-14T09:48:04.350597+0200 INFO Test cosine_similarity: 0.8162
2024-10-14T09:48:04.350651+0200 INFO Test loss: -0.3639
2024-10-14T09:48:04.350690+0200 INFO Test mean_absolute_error: 0.0017
2024-10-14T09:48:04.350726+0200 INFO Test mean_squared_error: 0.0000
2024-10-14T09:48:04.350766+0200 INFO Test pearson_correlation: 0.4482
2024-10-14T09:48:04.350800+0200 INFO Test pearson_correlation_log: 0.6173
2024-10-14T09:48:04.350832+0200 INFO Test zero_penalty_metric: 29.8548
```

>  Do you have the feeling the model does not work at all if you do example predictions on random test set regions?

I have not tried because I did not want to be influenced but yes I will do the heatmaps

> Pseudobulks:

Indeed, I have some clusters with few cells:
```
   0    1    2    3    4    5    6    7    8 
1676 1564  707  676  279    0   89   63   62 
```

Maybe I should restrict only to the clusters 0 to 6, what do you think?

For the logger it seems that the 'wandb' option is not compatible with pytorch...

Thank you so much for all these inputs.

**Comment by nkempynck** [comment]
I think the combination of the lower number of regions, and the low amount of cells for around half of your clusters, is the main reason why the model is not performing optimally. As you propose, I would only take cluster 0, 1, 2, 3 and 4. For the others the signal to noise in the bigwigs will not be sufficient to learn from.

**Comment by LukasMahieu** [comment]
I believe they're training on counts, not peak height.
Either way, very high cosine similarity and very low MSE/MAE makes me think your data is very sparse (which is not necessarily a problem).
Check the distributions of your data with crested.pl.hist.distribution.

Keep in mind that this is a deep learning tool. Like with any deep learning model trained on any type of data, there is no guarantee it will behave exactly as expected from the get go. We also need to experiment with different model parameters, preprocessing steps, ... to get to a model we're satisfied with. There's no known way to success.

Generally, if the model does not train at all, it's usually a data quality issue, but I agree with Niklas that these metrics don't look all that bad. You need to do downstream analyses to be sure if it's (not) working.

**Comment by lldelisle** [comment]
For your info, my input is:
```python
    adata = crested.import_bigwigs(
        bigwigs_folder="wt_Fate_bw_selected",
        regions_file="wt_Fate_all_peaks.bed",
        chromsizes_file=chromsizes_file,
        target_region_width=1000,  # optionally, use a different width than the consensus regions file (500bp) for the .X values calculation
        target="mean",  # or "max", "count", "logcount" --> what we will be predicting
    )
```

**Comment by lldelisle** [comment]
Here is the correlation plot in the cluster with the highest correlation:
![image](https://github.com/user-attachments/assets/6bf31768-93f6-4050-8626-176ee684cfbc)

All values are super low.

In addition, here is the self-correlation heatmap:
![image](https://github.com/user-attachments/assets/6fd88d80-4968-4911-b67b-afd50482c804)
and the ground truth/prediction heatmap:
![image](https://github.com/user-attachments/assets/09f0b356-30e3-4b4a-917c-04280f47fd55)

I guess this may be my cell types are not different enough...

Here is a plot on my locus of interest on my cell type of interest:
![image](https://github.com/user-attachments/assets/72c5829c-7d2b-415c-9a3c-390c076d2c2e)

**Comment by lldelisle** [comment]
Sorry for this long absence. I reran the modeling with a 20x to be in the 0-2 range as @nkempynck suggested and here are the results:
```
# New model with 20X:
2024-11-27T08:37:40.723681+0100 INFO Test concordance_correlation_coefficient: 0.3744
2024-11-27T08:37:40.723874+0100 INFO Test cosine_similarity: 0.9447
2024-11-27T08:37:40.723927+0100 INFO Test loss: -0.0936
2024-11-27T08:37:40.723967+0100 INFO Test mean_absolute_error: 0.0398
2024-11-27T08:37:40.723998+0100 INFO Test mean_squared_error: 0.0069
2024-11-27T08:37:40.724032+0100 INFO Test pearson_correlation: 0.6127
2024-11-27T08:37:40.724074+0100 INFO Test pearson_correlation_log: 0.6794
2024-11-27T08:37:40.724106+0100 INFO Test zero_penalty_metric: 1.6304
# Old model:
2024-11-27T08:37:06.442478+0100 INFO Test concordance_correlation_coefficient: 0.4440
2024-11-27T08:37:06.442623+0100 INFO Test cosine_similarity: 0.9282
2024-11-27T08:37:06.442683+0100 INFO Test loss: -0.5646
2024-11-27T08:37:06.442731+0100 INFO Test mean_absolute_error: 0.0020
2024-11-27T08:37:06.442768+0100 INFO Test mean_squared_error: 0.0000
2024-11-27T08:37:06.442802+0100 INFO Test pearson_correlation: 0.5481
2024-11-27T08:37:06.442834+0100 INFO Test pearson_correlation_log: 0.6546
2024-11-27T08:37:06.442864+0100 INFO Test zero_penalty_metric: 0.3717
```

The predictions on my locus looks similar:
![image](https://github.com/user-attachments/assets/d710295c-5e0f-4f46-8dcc-acabcddf7d75)

The issue is that one of our enhancer of interest in which we wanted to get the potential TF binding sites is not well predicted by the model.
